gpii/node_modules/ontologyHandler/src/ontologyHandler.js

```diff
@@ -108,6 +124,165 @@ URLs - this aspect is taken care of in the preferences server component.
+            edges.push({
+                source: mapping.target,
+                target: mapping.source
+            });
+        });
+        that.options.ontologyTransformRoutes =
+            gpii.ontologyHandler.floydWarshall.fullAlgorithmWithReconstruction(edges);
+    };
+
+    /**
+     * Implementation of the Floyd-Warshall algorithm to determine the shortest
+     * paths between a set of named edges. This is mostly generic, but currently
+     * used in generating a path between multiple linked ontologies.
+     * https://en.wikipedia.org/wiki/Floyd–Warshall_algorithm#Path_reconstruction
+     * Currently assumes that all edges are weight 1.
+     *
+     * @edges (Array) - An array of arrays of length 2, consisting of the named
+     *     edges such as [['A','B'], ['B','C']]
+     * @returns (Object) - An object containing the necessary structures to resolve
+     *     paths using the floydWarshallPath function. Contains 4 members:
+     *     vertices (Object) key/value pair where key is the vertex name, and value
+     *         is an integer to be used in the algorithm marking it as a vertex.
+     *         For the ontology path use case, the vertex is the name of the
+     *         ontology.
+     *     numVertices (Integer) Convenience variable for the total number of vertices(ontologies).
+     *     dist (2D Array) dist and next are matrices used in the algorithm, see
+     *         documentation link above for algorithm details.
+     *     next (2D Array) see dist above
+     */
+    gpii.ontologyHandler.floydWarshall.fullAlgorithmWithReconstruction = function ( edges ) {
+        var data = {};
+        gpii.ontologyHandler.floydWarshall.algorithmInitVertices(data, edges);
+        gpii.ontologyHandler.floydWarshall.algorithmInitMatrices(data, edges);
+        gpii.ontologyHandler.floydWarshall.algorithmCore(data.dist, data.next, data.numVertices);
+        return data;
+    };
+
+    /**
+     * Internal Use only
+     * Function used in gpii.ontologyHandler.floydWarshall.fullAlgorithmWithReconstruction
+     * to initialize the mapping of vertices named by ontology to integers used in
+     * the algorithm.
+     *
+     * @data (Object) Data structure used in main function.
+     * @edges (Array) List of edges where each is an Object contains a directed source and
+     *      target parameter.
+     * @returns Does not a return a value, operates directly on the data structure.
+     */
+    gpii.ontologyHandler.floydWarshall.algorithmInitVertices = function (data, edges) {
+        data.vertices = {};
+        fluid.each(edges, function (edge) {
+            gpii.ontologyHandler.floydWarshall.incrementingHashAppend(data.vertices, edge.source);
+            gpii.ontologyHandler.floydWarshall.incrementingHashAppend(data.vertices, edge.target);
+        });
+        // Assign each vertex an integer key for the lookup arrays
+        fluid.each(Object.keys(data.vertices), function (vert, i) {
```

Comment:
Better to hoist out the repeated use of Object.keys from this function and the one it calls into a simple field. They should both operate on a structure record: {
    hash: data.vertices,
    count: 0
}